drivers/pca9633/pca9633.c

```diff
@@ -0,0 +1,281 @@
+/*
+ * Copyright (C) 2020 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_pca9633
+ * @brief       Device driver for the PCA9633 I2C PWM controller
+ * @author      Hendrik van Essen <hendrik.ve@fu-berlin.de>
+ * @file
+ * @{
+ */
+
+#include <stdio.h>
+
+#include "pca9633.h"
+#include "pca9633_regs.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+/**
+ * @brief Write data to a register.
+ *
+ * @param[in] dev       Device descriptor of the PCA9633
+ * @param[in] reg       Register address to write to
+ * @param[in] data      Data to write
+ *
+ * @return  0 on success
+ * @return  -PCA9633_ERROR_I2C if acquiring of I2C bus fails
+ * @return  -EIO When slave device doesn't ACK the byte
+ * @return  -ENXIO When no devices respond on the address sent on the bus
+ * @return  -ETIMEDOUT When timeout occurs before device's response
+ * @return  -EINVAL When an invalid argument is given
+ * @return  -EOPNOTSUPP When MCU driver doesn't support the flag operation
+ * @return  -EAGAIN When a lost bus arbitration occurs
+ */
+static int _write_reg(pca9633_t* dev, uint8_t reg, uint8_t data);
+
+/**
+ * @brief Read data from a register.
+ *
+ * @param[in] dev       Device descriptor of the PCA9633
+ * @param[in] reg       Register address to read from
+ * @param[out] data     Byte read from given registerAddress
+ *
+ * @return  0 on success
+ * @return  -PCA9633_ERROR_I2C if acquiring of I2C bus fails
+ * @return  -EIO When slave device doesn't ACK the byte
+ * @return  -ENXIO When no devices respond on the address sent on the bus
+ * @return  -ETIMEDOUT When timeout occurs before device's response
+ * @return  -EINVAL When an invalid argument is given
+ * @return  -EOPNOTSUPP When MCU driver doesn't support the flag operation
+ * @return  -EAGAIN When a lost bus arbitration occurs
+ */
+static int _read_reg(pca9633_t* dev, uint8_t reg, uint8_t* data);
+
+int pca9633_init(pca9633_t *dev, const pca9633_params_t *params)
+{
+    assert(dev);
+    assert(params);
+
+    dev->params = *params;
+
+    i2c_init(dev->params.i2c_dev);
+
+    int rc = _write_reg(dev, PCA9633_REG_MODE1, 0x0);
+    _write_reg(dev, PCA9633_REG_MODE2, 0x0);
+
+    if (rc != PCA9633_OK) {
+        return rc;
+    }
+
+    pca9633_set_ldr_state_all(dev, PCA9633_LDR_STATE_IND_GRP);
+    pca9633_set_group_control_mode(dev, PCA9633_GROUP_CONTROL_MODE_DIMMING);
+    pca9633_set_rgb(dev, 255, 255, 255);
```

Comment:
Does that enable the output on init?
I think that's not what most users would want - only enable the output when the user has requested it.

Reply:
It's turned on by default. The lines
```
pca9633_set_ldr_state_all(dev, PCA9633_LDR_STATE_IND_GRP);
pca9633_set_group_control_mode(dev, PCA9633_GROUP_CONTROL_MODE_DIMMING);
```
are somewhat of an initialization, so they should be kept. But I could remove
```
pca9633_set_rgb(dev, 255, 255, 255);
```
and exchange it with a call of `pca9633_turn_off()` if that's how the drivers normally operate.

Reply:
Yea you don't want the LED to flash each time the device is turned on.
And one could connect something else entirely to those PWM outputs, like a motor or such.
Generally, you don't want something to happen without the user explicitly calling a function to make it happen.

Reply:
Ah I mixed this in my head with another driver here. I had another look to the data sheet. For this one, there is no global `turned off` state. All LDRs (4 different for the pwm channels) are set separatly (`PCA9633_LDR_STATE_OFF` by default).

Because we do `pca9633_set_group_control_mode(dev, PCA9633_GROUP_CONTROL_MODE_DIMMING);` we don't need `pca9633_set_group_control_mode(dev, PCA9633_GROUP_CONTROL_MODE_DIMMING);`

Long story short: I remove all three lines and no need for `pca9633_turn_off` :D